openwebwork/webwork2
# UTF8 Support

This pull requests adds better support for UTF8 characters, from the rendered webpage all the to the database.  While it looks simple its got a surprising number of ramifications and will need a lot of testing. 

To install:  Checkout both this pull request _and_ pull request openwebwork/pg#278

To test: 
- [ ] Check that none of the existing functionality is broken. 
- [ ] Check that you can use the Russian language setting without long character warnings and that it looks correct.  (This will test long character utf8 coming from the localization engine). 
- [ ] Check that you can use utf8 in anything that writes or reads from a file. A good string to test with is "á, é, í, ó, ú, ü, ñ, ¿, ¡ 한글 漢字" This will include
  -  The site_info box
  - The course_info box
  - Email files (as well as the grade message feature)
  - Saving and editing pg files. 
  - Rendering pg files, both via problem pages and xmlrpc (although I make no promises about unicode correctly passing through macros)
- [ ] Check that you can use "short" utf characters in text fields (like first name, last name, comment, set description, etc...)  You can use "á, é, í, ó, ú, ü, ñ" as a test string. 
- [ ] Check that you can use "long" utf characters if you set up your mysql server to support it.  
  -  Log into your mysql server and run the following to change the default character set of webwork to utf8
  
  ```
  alter database webwork character set utf8 collate utf8_general_ci;
  ```
  - Create a new course.  (Then maybe double check that the tables created have a default character set of utf8).  Then check that you can save longer utf8 characters in text fields.  You can test with " á, é, í, ó, ú, ü, ñ, ¿, ¡ 한글 漢字"
- [ ] Check that achievements (and other features which use nfreeze) still work.  Note:  I changed achievements to use a base64 version of freeze and thaw.  This will be compatible with older installations, but the functionality should be checked.
